cli/loader_lib.c

```diff
@@ -153,6 +152,206 @@ JL_DLLEXPORT const char * jl_get_libdir()
+        len -= bytes_written;
+        written_sofar += bytes_written;
+    }
+}
+
+// Where the buf passed in was malloced
+static void read_wrapper(int fd, char **ret, size_t *ret_len)
+{
+    // Allocate an initial buffer
+    size_t len = JL_PATH_MAX;
+    char *buf = (char *)malloc(len + 1);
+    if (!buf) {
+        fprintf(stderr, "Error during libstdcxxprobe in parent process:\n");
+        fprintf(stderr, "malloc() returned NULL.\n");
+        exit(1);
+    }
+
+    // Read into it, reallocating as necessary
+    size_t have_read = 0;
+    while (1) {
+        ssize_t n = read(fd, buf + have_read, len - have_read);
+        have_read += n;
+        if (n == 0) break;
+        if (n == -1 && errno != EINTR) {
+            fputs("Error during libstdcxxprobe in parent process:", stderr);
+            perror("read");
+            exit(1);
+        }
+        if (n == -1 && errno == EINTR) continue;
+        if (have_read == len) {
+            buf = (char *)realloc(buf, 1 + (len *= 2));
+            if (!buf) {
+                fputs("Error during libstdcxxprobe in parent process:", stderr);
+                fprintf(stderr, "malloc() returned NULL.\n");
+                exit(1);
+            }
+        }
+    }
+
+    *ret = buf;
+    *ret_len = have_read;
+}
+
+// On Linux, it can happen that the system has a newer libstdc++ than the one we ship,
+// which can break loading of some system libraries: <https://github.com/JuliaLang/julia/issues/34276>.
+
+// Return the path to the libstdcxx to load. Otherwise, print the error and exit.
+// The path returned must be freed.
+static char *libstdcxxprobe(void)
+{
+    // Create the pipe and child process.
+    int fork_pipe[2];
+    int ret = pipe(fork_pipe);
+    if (ret == -1) {
+        perror("Error during libstdcxxprobe:\npipe");
+        exit(1);
+    }
+    // Make sure the buffers are clean before forking.
+    fflush(stdout);
+    fflush(stderr);
+    pid_t pid = fork();
+    if (pid == -1)  {
+        perror("Error during libstdcxxprobe:\nfork");
+        exit(1);
+    }
+    if (pid == (pid_t) 0) { // Child process.
+        ret = close(fork_pipe[0]);
+        if (ret == -1) {
+            perror("Error during libstdcxxprobe in child process:\nclose");
+            _exit(1);
+        }
+
+        // Open the first available libstdc++.so.
+        // If it can't be found, report so by exiting zero.
+        void *handle = dlopen("libstdc++.so", RTLD_LAZY);
+        if (!handle) _exit(0);
+
+        // See if the version is compatible
+        char *dlerr = dlerror(); // clear out dlerror
+        void *sym = dlsym(handle, GLIBCXX_LEAST_VERSION);
+        dlerr = dlerror();
+        if (dlerr) {
+            // We can't use the library that was found, use the default one instead.
+            char *bundled;
+            size_t len;
+            get_bundled_folder(&bundled, &len);
+            strcat(bundled, "libstdc++.so");
+
+            write_wrapper(fork_pipe[1], bundled, len);
```

Comment:
since the parent already is required to implement this, I don't think we need to reimplement it here
```suggestion
```

Reply:
I too have been annoyed at the code duplication. Originally I checked to see if nothing was written over the pipe, and failed out of the function. I was then asked to change it, but I think it works better.